### DISCUSSION TITLE
How to have an alias in a container?

### DISCUSSION THREAD

**Discussion by VladimirFokow** [hpc/charliecloud#1854] [General]
I have created an image. There is an executable in it, for example: `/path/to/executable.sh`

When I `ch-run` and enter the container's bash, I want to have an easy alias for it, for example: `e`

With normal linux, I would add this to `~/.bashrc`:
```bash
alias e='/path/to/executable.sh'
```

But I'm having a very hard time figuring out how to do it with charliecloud... 
Is it even possible? 

<br /><br />

<details>
<summary> What I've tried </summary>
<br />

- I have created an executable `entryscript.sh` and copied it into the image (when building from `Dockerfile`) to `/entryscript.sh`.  
  It contains: 
  ```bash
    #!/bin/bash

    echo "Inside of entryscript"
    alias e='/path/to/executable.sh'
   ```

- I'm running `ch-run ... -- `  and I'm trying different commands after `--`,  none of which gave me a shell inside of the running container with the `e` alias available:
  - `bash -c "source /entryscript.sh"`
  - `bash && source /entryscript.sh`
  - `bash -c "alias e='/path/to/executable.sh' && bash"`
  - `bash -c "alias e='/path/to/executable.sh'" && bash`
  - `bash && alias e='/path/to/executable.sh'`

<br />

I don't really know what to try next.. I would appreciate any help!
(Would this be easier to achieve if `CMD` or `ENTRYPOINT` Dockerfile instructions were available?)

<br /><br />

---

A possible workaround:

I  couldn't make aliases work - instead, can use functions for this specific use-case.

In `entryscript.sh` have:
```bash
#!/bin/bash

e() {
    /path/to/executable.sh "$@"
}
export -f e

/bin/bash
```

and then

```bash
ch-run ... -- bash /entryscript.sh
```

But instead, I'd like to know if a general approach exists - an alternative of sourcing `~/.bashrc` - automatically when starting a container.
The current workaround may not emulate all desired behaviors exactly.

</details>

<br />

**Comment by reidpr** [comment]
The first thing I'd suggest is to embrace the principle that in containers, you don't need to share. That is, there is no need to adjust your shell environment; rather, you can just manipulate the image to your and your application's specialized liking.

In this case, it sounds to me like you have a Dockerfile that looks something like this:

```
FROM foo
COPY ./executable.sh /path/to/executable.sh
```

and you want to be able to:

```
$ ch-run ... -- executable.sh
```

If that's correct, I'd suggest either:

```
FROM foo
COPY ./executable.sh /usr/local/bin
```

(or whatever's already in your default path, e.g. `/usr/bin`)

or:

```
FROM foo
COPY ./executable.sh /path/to/executable.sh
RUN ln -s /path/to/executable.sh /usr/local/bin
```

We do have an issue to support `ENTRYPOINT` (#780) and hopefully that will be available soon.

Does that help?

**Reply by VladimirFokow** [reply]
hello,

I've tried `ln -s`, there are some errors due to the wrong path.
I can't move the executable to another location - it has hard-coded relative paths inside.

I don't want to run only it; I'd like to enter an interactive bash session, and have access to it and other utilities via their short names...

----

I didn't really understand "in containers, you don't need to share. That is, there is no need to adjust your shell environment"

Isn't an interactive bash session a valid use case? With the convenience of aliases

**Reply by VladimirFokow** [reply]
Are aliases (and more generally, sourcing some file upon starting a container) possible now,
or will it be possible in any future (would it be possible even with ENTRYPOINT)?

**Reply by reidpr** [reply]
The "don't need to share" means that you can put whatever you want at standard paths and don't have to worry about messing up other users by doing so. An interactive bash session is definitely a valid use case; the suggestion is to make that work without shell configuration. However, it sounds like your executable doesn't support that.

It does look like there are ways to have Bash source an arbitrary script at start-up ([docs](https://www.gnu.org/software/bash/manual/html_node/Invoking-Bash.html)). For example, I *think*:

```
$ ch-run ... -- bash --rcfile /my/aliases.sh
```

will give you an interactive shell with `/my/aliases.sh` sourced first. Or:

```
$ ch-run ... -- bash --login
```

will give you an interactive *login* shell and so it will source `/etc/profile`. Finally, by default `$HOME` is `/root`, so this may give an interactive shell that sources `/root/.bashrc`:

```
$ ch-run ... -- bash
```

Finally, another option is `ch-run --home`, which binds your home directory into the container, so interactive Bash shell would then source your normal `.bashrc`.

Possibly this is leading towards a new FAQ entry?

**Reply by VladimirFokow** [reply]
oh, thank you! 

 `bash --rcfile` does exactly what I need, and you've provided great other alternatives as well!
